lib/Dialect/Calyx/Transforms/AffineParallelUnroll.cpp

```diff
@@ -33,6 +36,181 @@ using namespace mlir;
+  }
+};
+} // namespace llvm
+
+DenseMap<Operation *, SmallVector<int64_t, 4>>
+CalyxParLICM::computeConstMemAccessIndices(AffineParallelOp affineParallelOp) {
+  DenseMap<Operation *, SmallVector<int64_t, 4>> constantMemAccessIndices;
+
+  MLIRContext *ctx = affineParallelOp->getContext();
+  auto executeRegionOps =
+      affineParallelOp.getBody()->getOps<scf::ExecuteRegionOp>();
+  for (auto executeRegionOp : executeRegionOps) {
+    executeRegionOp.walk([&](Operation *op) {
+      if (!isa<AffineReadOpInterface, AffineWriteOpInterface>(op))
+        return WalkResult::advance();
+
+      auto read = dyn_cast<AffineReadOpInterface>(op);
+      AffineMap map = read ? read.getAffineMap()
+                           : cast<AffineWriteOpInterface>(op).getAffineMap();
+      ValueRange mapOperands =
+          read ? read.getMapOperands()
+               : cast<AffineWriteOpInterface>(op).getMapOperands();
+
+      SmallVector<Attribute> operandConsts;
+      for (Value operand : mapOperands) {
+        if (auto constOp =
+                operand.template getDefiningOp<arith::ConstantIndexOp>()) {
+          operandConsts.push_back(
+              IntegerAttr::get(IndexType::get(ctx), constOp.value()));
+        } else {
+          return WalkResult::advance();
+        }
+      }
+
+      SmallVector<int64_t, 4> evaluatedIndices, foldedResults;
+      bool hasPoison = false;
+      map.partialConstantFold(operandConsts, &foldedResults, &hasPoison);
+      if (!(hasPoison || foldedResults.empty()))
+        constantMemAccessIndices[op] = foldedResults;
+
+      return WalkResult::advance();
+    });
+  }
+
+  return constantMemAccessIndices;
+}
+
+LogicalResult CalyxParLICM::writeOpAnalysis(
+    DenseMap<Operation *, SmallVector<int64_t, 4>> &constantMemAccessIndices) {
+  for (auto &[memOp, constIndices] : constantMemAccessIndices) {
+    auto writeOp = dyn_cast<AffineWriteOpInterface>(memOp);
+    if (!writeOp) {
+      continue;
+    }
+
+    auto parentExecuteRegionOp =
+        writeOp->getParentOfType<scf::ExecuteRegionOp>();
+    auto key = std::pair(writeOp.getMemRef(), constIndices);
+    if (constantWriteOpIndices.find(key) != constantWriteOpIndices.end() &&
+        constantWriteOpIndices[key] != parentExecuteRegionOp) {
+      // It cannot occur more than once in different parallel regions, which
+      // will result in write contention.
+      return failure();
+    }
+
+    constantWriteOpIndices[key] = parentExecuteRegionOp;
+  }
+  return success();
+}
+
+void CalyxParLICM::readOpHoistAnalysis(
+    AffineParallelOp affineParallelOp,
+    DenseMap<Operation *, SmallVector<int64_t, 4>> &constantMemAccessIndices) {
+  for (auto &[memOp, constIndices] : constantMemAccessIndices) {
+    auto readOp = dyn_cast<AffineReadOpInterface>(memOp);
+    if (!readOp)
+      continue;
+
+    auto key = std::pair(readOp.getMemRef(), constIndices);
+    if (constantWriteOpIndices.contains(key))
+      continue;
+    constantReadOpCounts[key]++;
```

Comment:
Would something like this work to avoid double lookup?
```
auto it = constantWriteOpIndices.find(key);
if (it == constantWriteOpIndices.end())
  continue;
++it->second;
```